### PR TITLE
[SD] Update all the utilities to make web and CLI codebase closer

### PR DESCRIPTION
**Pull request by Shukla-Gaurav**
At this point, all the utilities of SD web and CLI are exactly same. 
Also the `web/models/stable_diffusion/main.py` can be directly executed.

Signed-Off-by: Gaurav Shukla <gaurav@nod-labs.com>

### CODE REVIEW

**Comment by Shukla-Gaurav** [comment]
@qedawkins Can you please help me test this patch? 
This will help a lot to merge the CLI and web.

**Comment by Shukla-Gaurav** [comment]
@powderluv Once tested without any perf regression, the web folder will contain the cli as well.

**Comment by Shukla-Gaurav** [comment]
Will test it tomorrow thoroughly before merging.

File: web/models/stable_diffusion/cache_objects.py [hunk 1]
```diff
@@ -43,11 +47,20 @@
...
+if args.version == "v2_1base":
+    schedulers[
+        "SharkEulerDiscrete"
+    ] = SharkEulerDiscreteScheduler.from_pretrained(
+        model_config[args.version],
+        subfolder="scheduler",
+    )
+    schedulers["SharkEulerDiscrete"].compile()
+else:
+    schedulers["SharkEulerDiscrete"] = EulerDiscreteScheduler.from_pretrained(
```
Comment (qedawkins) [inline web/models/stable_diffusion/cache_objects.py:59]: I don't see why the SharkEulerDiscrete scheduler is exclusive to v2.1base.

File: web/models/stable_diffusion/main.py [hunk 2]
```diff
@@ -10,6 +10,24 @@
 from random import randint
 import numpy as np
 import time
+import sys
+
+
+# Helper function to profile the vulkan device.
+def start_profiling(file_path="foo.rdc", profiling_mode="queue"):
```
Comment (qedawkins) [inline web/models/stable_diffusion/main.py:17]: We should check whether anyone is using this feature and if not, drop it. Another option would be to make it available for all models via SharkRunner.

File: web/models/stable_diffusion/utils.py [hunk 3]
```diff
@@ -168,23 +167,27 @@ def map_device_to_name_path(device, key_combination=3):
...
+            f"Found device {device_name}. Using target triple {args.iree_vulkan_target_triple}."
+        )
+
+        # use tuned models only in the case of stablediffusion/fp16 and rdna3 cards.
+        if (
+            args.variant != "stablediffusion"
+            or args.precision != "fp16"
+            or "rdna3" not in args.iree_vulkan_target_triple
+        ):
             args.use_tuned = False
```
Comment (qedawkins) [inline web/models/stable_diffusion/utils.py:187]: nit: add a warning/log similar to the one added below if use_tuned was set to true.

File: web/models/stable_diffusion/utils.py [hunk 4]
```diff
@@ -168,23 +167,27 @@ def map_device_to_name_path(device, key_combination=3):
...
 
         if args.use_tuned:
             print("Using tuned models for rdna3 card")
     else:
-        if args.use_tuned:
-            print("Tuned models not currently supported for device")
-            args.use_tuned = False
+        args.use_tuned = False
+        print("Tuned models are currently not supported for this setting.")
```
Comment (qedawkins) [inline web/models/stable_diffusion/utils.py:193]: I think the message here is still only necessary if use_tuned is true.

**Review by qedawkins — CHANGES_REQUESTED** [review]
Still doing some e2e verifications, but I have a few comments first.

File: web/models/stable_diffusion/main.py [hunk 5]
```diff
@@ -162,3 +184,25 @@ def stable_diff_inf(
     text_output += f"\nTotal image generation time: {total_time:.4f}sec"
 
     return pil_images[0], text_output
+
+
+if __name__ == "__main__":
```
Comment (qedawkins) [inline web/models/stable_diffusion/main.py:189]: This won't work due to two different entry points with different relative imports (index.py and models/stable_diffusion/main.py)
```
Traceback (most recent call last):
  File "/home/quinn/nod/SHARK/web/models/stable_diffusion/main.py", line 5, in <module>
    from models.stable_diffusion.cache_objects import (
ModuleNotFoundError: No module named 'models'
```
This is why we need #694 first to enable absolute imports for all of the relevant SD code.

**Review by qedawkins — CHANGES_REQUESTED** [review]

File: web/models/stable_diffusion/main.py (hunk 5 above)
Comment (Shukla-Gaurav) [reply web/models/stable_diffusion/main.py:189]: @qedawkins Let's land #694 first then. Thanks!

File: web/models/stable_diffusion/cache_objects.py (hunk 1 above)
Comment (Shukla-Gaurav) [reply web/models/stable_diffusion/cache_objects.py]: Fixed. Thanks!

File: web/models/stable_diffusion/utils.py (hunk 3 above)
Comment (Shukla-Gaurav) [reply web/models/stable_diffusion/utils.py:189]: Added same message.

File: web/models/stable_diffusion/main.py (hunk 5 above)
Comment (Shukla-Gaurav) [reply web/models/stable_diffusion/main.py]: Ok, so I have removed this second entry point from this PR as it is already there in #694. Now this PR is good to go before #694. I will take a look at the merging stuff with #694.

File: web/models/stable_diffusion/main.py (hunk 2 above)
Comment (Shukla-Gaurav) [reply web/models/stable_diffusion/main.py:18]: yeah, we can discuss on this and put it at the right place.

**Comment by Shukla-Gaurav** [comment]
@PhaneeshB @qedawkins I have tested it all and this is good to go. After this I will look into the movement PR #694.

**Review by PhaneeshB — APPROVED** [review]
LGTM

**Review by powderluv — APPROVED** [review]
Please resolve all comments. Please also generate an exe and validate the models work with the exe and then land

**Comment by Shukla-Gaurav** [comment]
> Please resolve all comments. Please also generate an exe and validate the models work with the exe and then land

all comments are addressed and .exe is working fine.

**Pull request merged**